kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -395,20 +396,86 @@ private static void validateSchemaEvolution(
     }
   }
 
-  /* Validates a given schema evolution by using field ID
-  as the source of truth for identifying fields */
+  /**
+   * Validates a given schema evolution by using field ID as the source of truth for identifying
+   * fields
+   */
   private static void validateSchemaEvolutionById(
       StructType currentSchema, StructType newSchema, Metadata metadata) {
     Map<Long, StructField> currentFieldsById = fieldsById(currentSchema);
     Map<Long, StructField> updatedFieldsById = fieldsById(newSchema);
     SchemaChanges schemaChanges = computeSchemaChangesById(currentFieldsById, updatedFieldsById);
-
-    // Validate physical name consistency on the updated fields
     validatePhysicalNameConsistency(schemaChanges.updatedFields());
+    validateUpdatedSchemaCompatibility(schemaChanges);
+    // ToDo Potentially validate IcebergCompatV2 nested IDs
+  }
+
+  /**
+   * Verifies the following
+   *
+   * <ul>
+   *   <li>no non-nullable fields are added
+   *   <li>no tightening of existing fields
+   *   <li>no type changes are performed
+   */
+  private static void validateUpdatedSchemaCompatibility(SchemaChanges schemaChanges) {
+    for (StructField addedField : schemaChanges.addedFields()) {
+      if (!addedField.isNullable()) {
+        throw new KernelException(
+            String.format("Cannot add non-nullable field %s", addedField.getName()));
+      }
+    }
 
-    // ToDo Validate IcebergCompatV2 nested IDs
+    for (Tuple2<StructField, StructField> updatedFields : schemaChanges.updatedFields()) {
+      validateNoTypeChange(updatedFields._1, updatedFields._2);
+    }
+  }
+
+  /**
+   * Validate that there was no change in type from existing field from new field, excluding dropped
+   * columns for structs. Note, fields are resolved by case-sensitive name
+   */
+  private static void validateNoTypeChange(StructField existingField, StructField newField) {
+    if (existingField.isNullable() && !newField.isNullable()) {
+      throw new KernelException(
+          String.format(
+              "Cannot tighten the nullability of existing field %s", existingField.getName()));
+    }
 
-    // ToDo: Validate schema compatibility
+    // Both fields are structs, ensure there's no changes in the individual fields
+    if (existingField.getDataType() instanceof StructType
+        && newField.getDataType() instanceof StructType) {
+      StructType existingStruct = (StructType) existingField.getDataType();
+      Map<String, List<StructField>> existingFieldsByName =
+          existingStruct.fields().stream().collect(Collectors.groupingBy(StructField::getName));
+      StructType newStruct = (StructType) newField.getDataType();
+      for (StructField newNestedField : newStruct.fields()) {
+        List<StructField> existingNestedField = existingFieldsByName.get(newNestedField.getName());
+        // The existing field is guaranteed to be well-defined, there should only be one field with
+        // the name
```

Comment:
Double check this, the passed in values are already based on ID so not sure why this was origianlly added